### ISSUE TITLE
Build System

### ISSUE THREAD

**Issue by benbjohnson** [flynn/flynn#1865]
## Overview ![status](https://img.shields.io/badge/status-draft-red.svg)

This is a proposal for a new build system in Flynn that will function as both the build tool for applications running in Flynn as well as the build tool for the Flynn project itself. There are other existing build tools but there are none (to my knowledge) that are written in Go and that can be easily integrated into the Flynn project's workflow.

There is currently no name for this project, however in the tradition of other `make`-like tools, it seems like `flake` (`flynn` + `make`) would be a reasonable choice.
## Goals

There are 3 goals of the build system:
1. Functional: Outputs from a package should only depend on its inputs. Given the same inputs, the output should always be the exact same.
2. Content Addressable: A file should be addressable in a particular state based on its contents and permissions. This means that output files can be cached in order to perform incremental builds. Changes to a file's contents should trigger a rebuild of dependencies.
3. Pluggable: Allow the build system to plug into a local or remote Flynn cluster to allow parallel builds and utilize a shared build cache.
### Dependency graph

Flynn's current build system is [`tup`](http://gittup.org/tup/). Tup creates a [DAG](https://en.wikipedia.org/wiki/Directed_acyclic_graph) of dependencies by using a [FUSE file system](http://fuse.sourceforge.net/) to monitor file access by subprocesses. This differs from traditional `make` in that input files do not need to be specified explicitly. The downside is that the FUSE filesystem can be finicky and requires set up. Another downside of tup is that its rule language is unfamiliar to developers coming from a build system like `make`.

More details of tup's dependency graph can be found in [Build System Rules & Algorithms](http://gittup.org/tup/build_system_rules_and_algorithms.pdf).

A similar approach to dependency graph management will be used in the new build system, however, we would like to avoid FUSE if possible.
#### FUSE Alternatives

Some alternatives to FUSE have been evaluated:
- [`seccomp(2)`](http://man7.org/linux/man-pages/man2/seccomp.2.html) - The Secure Computing state allows syscalls to be filtered and trapped. This would allow us to capture `read` calls from subprocesses within the build and rewrite them as needed. I spent some time trying to implement this but rewriting syscalls is a fairly obscure feature and I haven't had luck with it so far. I attempted to use [`libseccomp-golang`](https://github.com/seccomp/libseccomp-golang) but there's not much documentation so I didn't get far with it. `seccomp(2)` is also only available on later versions of Linux so this would not be compatible with Mac OS X or Windows.
- [`ptrace(2)`](http://linux.die.net/man/2/ptrace) - This mechanism would let us trap syscalls from subprocesses similar to how a debugger would trace a program. It would let us build the dependency graph without using a mounted file system and it is cross compatible with Mac OS X. It would let us pause subprocesses in order to pull in dependencies, however, I think our ability to rewrite syscalls is limited.
- [`LD_PRELOAD`](http://man7.org/linux/man-pages/man8/ld.so.8.html) - This mechanism would let us preload our own implementations of standard library functions into child processes. With the ability to do shared libraries in Go 1.5 we could theoretically inject a shared library into subprocesses to rewrite `read()` calls in order to load dependencies. This is also available on Mac OS X as `DYLD_INSERT_LIBRARIES`.
### Rule language

Several build systems such as tup, [Bazel](http://bazel.io/), and [Ant](http://ant.apache.org/) use custom languages to define their build system. This provides extensibility for the build system, however, it also adds significant ramp up for the developer using the system. The Bazel documentation, for example, is nearly 200 pages when printed.

Alternatively there are build systems such as `make` and [Heroku buildpacks](https://devcenter.heroku.com/articles/buildpacks) which simply piggyback the underlying shell. This makes it very easy for users to get up and running with few surprises.

I suggest we go with the latter approach and provide a thin layer over the shell to make it easy to use. I think a Makefile-like syntax make sense. The build file would be split into two sections: imports & rules
#### Imports

The imports sections would simply define the dependencies of this build file in a URL format:

```
import "http://github.com/flynn/go#1.5.1" as go
import "http://github.com/flynn/flannel#9fa40e3" as flannel
```
#### Rules

Then rules would be able to utilize those dependencies in a `make`-style syntax:

```
bin/flynn-host: go flannel
        go build -o $@ ./host
```

The shell subprocess created for this target will be injected with one of the FUSE alternatives  (seccomp, ptrace, LD_PRELOAD) in order to determine the dependency graph. This will allow future runs to fetch external dependencies as needed and to only update parts of the dependency graph that have been changed.
#### Extensibility

`make` syntax can be limiting so adding a scripting language such as Lua is an option as well.

---

/cc @titanous

**Comment by titanous** [comment]
Overall this is great!

It's worth mentioning in the pluggable section that an explicit goal is to be able to point at a Flynn cluster (local or remote) and use it to execute the build in parallel and utilize a shared build cache.

I think we should evaluating Lua as an option for the build language, Tup has this available, and we do use it [for the CLI build](https://github.com/flynn/flynn/blob/master/cli/Tupfile.lua). Having a full programming language with well-defined semantics and the ability to create project-wide helpers/functions is really useful when building more complicated build rules. If possible I'd like to avoid designing a new build language, as maintaining it and learning it both can have significant overhead.

**Comment by benbjohnson** [comment]
I updated the description to reflect your notes. :+1:

**Comment by josephglanville** [comment]
:+1:  For Lua, it's easy to learn and easy to embed in anything (I assume it's no different for golang) it's also reasonably easy to sandbox Lua if needed.

**Comment by benbjohnson** [comment]
I've written systems that integrate with Lua before and I can't say that I love the language itself but it has two things going for it: 1) there are pure Go implementations, and 2) it's not JavaScript. :)

**Comment by lmars** [comment]
This sounds great!

Some questions:

> The imports sections would simply define the dependencies of this build file in a URL format

So are the imports just URLs to other build files whose outputs will somehow be exposed to this build (e.g. a tarball extracted into the build's workspace)?

Would we need to know the outputs of a rule like we do with `tup`, or would any file created by the rule be considered output (e.g. using overlayfs to track changes)? I'm thinking specifically about the example given in #1515; I don't really know upfront what the outputs of `bundle install` or `rake assets:precompile` are going to be.

Any thoughts on maintaining a cache? If my `Gemfile` changes, I need to re-run `bundle install`, but I would like that to run from the output of the previous `bundle install` so that gems are cached. Or do you think we would write a `Gemfile.lock` parser so that the build system is more aware of the dependencies?

Maybe these questions are too specific at this early stage, but I like to think in terms of a real example.

**Comment by benbjohnson** [comment]
> So are the imports just URLs to other build files whose outputs will somehow be exposed to this build (e.g. a tarball extracted into the build's workspace)?

We'll likely need some kind of build file to determine what to do with the output of a tarball if we need to do more than simply extract it. I was thinking that we could reference the files output by a build by using the build's name plus the filename (e.g. `go:bin/go`).

> Would we need to know the outputs of a rule like we do with tup, or would any file created by the rule be considered output.

We'll still need to specify outputs although if we're using Lua then there's flexibility with how that's done. Not every file output from the build step will be an asset we want to keep and use. I'm sure some build steps will have a bunch of temporary files (e.g. `.a` files).

> Any thoughts on maintaining a cache?

The output of the build steps will be content addressable we should only need to rebuild pieces that are invalidated by their dependencies changing. In the specific case of `bundle install`, I'm not quite sure how we'll try to integrate with bundler's gem caching since it has its own caching mechanism. Unless we can interface with bundler as a mock gem server but that sounds like a lot of work. Trying to rerun `bundle install` from the previous system state could work but I'm worried about reproducible builds.

**Comment by titanous** [comment]
Effectively implemented in #3817.
